### ISSUE TITLE
PPO Questions

### ISSUE THREAD

**Issue by natolambert** [lvwerra/trl#119]
I'm comparing the PPO implementation to the OpenAI one and the [implementation details blog post](https://iclr-blog-track.github.io/2022/03/25/ppo-implementation-details/) that goes through it. Wondering if some of these things improve performance. If not, it's good for understanding.

I'm guessing the discrepancy comes from the original vs learn to summarize work, which is interesting.

Some things to confirm:
- [ ] PPO update question: I was a little confused seeing `returns = advantages + values` l693, instead of `adv = returns - values` why did it end up like that?
- [ ] Some implementations use a [residual value prediction](https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/ppo2/model.py#L69) in clipping. Compared to [TRL](https://github.com/lvwerra/trl/blob/d2e8bcf8373726fb92d2110c500f7df6d0bd566d/trl/trainer/ppo_trainer.py#L717).

**Comment by natolambert** [comment]
It could be good to make things like this configurable in a branch and learning how these implementation details transfer to RLHF.

**Comment by DaehanKim** [comment]
imo, residual clipping seems beneficial to prevent policy loss spiking reported in #101 . It's probably coming from instability in value estimation.

**Comment by natolambert** [comment]
Yeah, I'm running residual clipping example(s), we'll see. At least it'll be good to have the option to try both.

**Comment by natolambert** [comment]
Residual value prediction didn't help with stability (it's crimson-dog)
<img width="973" alt="Screenshot 2023-01-30 at 1 41 28 PM" src="https://user-images.githubusercontent.com/10695622/215601730-c6a1593f-6b20-4b37-bced-623a963c6217.png">

**Comment by natolambert** [comment]
Also not a big help via the other approx KL formulation. W&B [here](https://wandb.ai/natolambert/TRL/runs/e1258rd6?workspace=user-natolambert). Though, it's slightly more stable? 

We'll see how this run finishes converging.
<img width="1422" alt="Screenshot 2023-01-30 at 2 33 23 PM" src="https://user-images.githubusercontent.com/10695622/215611270-74f8ff43-10cd-43dd-9eb0-db117b8af90d.png">
